### ISSUE TITLE
TensorFlow tf.data.Dataset API extremely slow for 3D-shaped pipeline

### ISSUE THREAD

**Issue by mimxrt** [tensorflow/tensorflow#37013]
**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.3 LTS
- TensorFlow installed from (source or binary): binary pip
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de 2.1.0
- Python version: Python 3.7.6
- CUDA/cuDNN version: 7.6.4.38-1
- GPU model and memory: Quadro RTX 6000 22752 MB

**Describe the current behavior**
In the data pipeline given below, the performance is extremely low. The pipeline produces output of the shape `((128, None, 128), (128, None, 5))` and if either dimension 0 (batching) or dimension 1 (windowing) is removed, the performance is much higher. Unfortunately both batching and windowing are required for efficient and meaningful training.

Additionally, the iteration of the Dataset does not work as expected with 3D output. The shape of the output tensor (loop body at the end: `print([x.shape for x in e])`) is not printed at all. When removing either windowing or batching, the shape is printed as expected. From this I suspect this issue could also be a bug.

Timings for comparison (note: number of time steps / dim 1 (`None`) is avg. 312.5 = 40000 / 128):
- 1x Shape `((128, None, 128), (128, None, 5))`:
**546.8976650238037 s**
- 1000x Shape `((128, 128), (128, 5))` (no windowing):
8.29618239402771 s * 312.5 / 1000 = **2.592556998133659375 s**
Speedup: 210x faster
- 1000x Shape `((None, 128), (None, 5))` (no batching):
65.43172574043274 * 128 / 1000 = **8,37526089477539072 s**
Speedup: 65x faster

**Describe the expected behavior**
It is expected for the performance to scale linearly with the amount of data. The shape of the output data should not have a big negative effect on performance.

**Standalone code to reproduce the issue** 
```
import time
import random
import numpy as np
import tensorflow as tf

print("TensorFlow {}".format(tf.__version__))

num_features = 128
num_labels = 5
batch_size = 128
win_size = 2048
len_max = 600000

# create some sample .tfrecord files with different length

l1 = len_max - 44260
X1, Y1 = (np.ones((l1, num_features + 2)), np.zeros((l1, num_labels)))
l2 = len_max - 121340
X2, Y2 = (np.ones((l2, num_features + 2)), np.zeros((l2, num_labels)))

def write_tfrecord(X, Y, fn):
    with tf.Graph().as_default():
        ds = tf.data.Dataset.from_tensor_slices((X, Y))

        sstring = ds.map(lambda *x: 
           tf.reshape(tf.py_function(lambda *v:
               tf.train.Example(features=tf.train.Features(feature={
                   "features": tf.train.Feature(float_list=tf.train.FloatList(value=v[0].numpy())),
                   "label": tf.train.Feature(float_list=tf.train.FloatList(value=v[1].numpy())),
               })).SerializeToString(), x, tf.string
           ), ())
        )

        writer = tf.data.experimental.TFRecordWriter(fn)
        writer_op = writer.write(sstring)

        sess = tf.compat.v1.Session()
        sess.run(tf.compat.v1.global_variables_initializer())
        sess.run(writer_op)
        sess.close()

files_base = ["./temp1.tfrecord", "./temp2.tfrecord"]

write_tfrecord(X1, Y1, files_base[0])
write_tfrecord(X2, Y2, files_base[1])

# take 200 random files as dataset

files = random.choices(files_base, k=200)

ds_fs = tf.data.Dataset.list_files(files, shuffle=True, seed=1)
fs_len = 0
for f in ds_fs:
    fs_len += 1
print("Reading {} tfrecord files...".format(fs_len))

# create a StaticHashTable holding the length for trimming

ds_f_len_table = tf.lookup.StaticHashTable(tf.lookup.KeyValueTensorInitializer(
        tf.constant(files_base),
        tf.constant([random.randint(30000, 50000), random.randint(30000, 50000)], dtype=tf.int64)
), -1)

# prepare the Dataset

def prep_ds_file(file):
    _ds = tf.data.TFRecordDataset(file)
    _ds = _ds.map(lambda x: tf.io.parse_single_example(x, {
        "features": tf.io.FixedLenFeature([num_features + 2], tf.float32),
        "label": tf.io.FixedLenFeature([num_labels], tf.float32),
    }), num_parallel_calls=tf.data.experimental.AUTOTUNE)
    print(_ds)

    _ds = _ds.flat_map(lambda v: tf.data.Dataset.from_tensors((v["features"][2:], v["label"])))
    print(_ds)

    _trunc = ds_f_len_table.lookup(file)
    _ds = _ds.take(_trunc)
    print(_ds)
    _num_tsteps = _trunc // batch_size

    ####################################################################################################
    # WINDOWING                                                                                        #
    ####################################################################################################
    _ds = _ds.window(size=_num_tsteps, shift=win_size//2, stride=1, drop_remainder=True)               #
    print(_ds)                                                                                         #
    _ds = _ds.flat_map(lambda x, y: tf.data.Dataset.zip((x.batch(_num_tsteps), y.batch(_num_tsteps)))) #
    print(_ds)                                                                                         #
    ####################################################################################################

    ##################################################
    # BATCHING                                       #
    ##################################################
    _ds = _ds.batch(batch_size, drop_remainder=True) #
    print(_ds)                                       #
    ##################################################

    return _ds


def prep_ds(files):
    _ds = files.flat_map(prep_ds_file)
    print(_ds)
    return _ds


ds = prep_ds(ds_fs)

# read/use the Dataset

ts = time.time()
for e in ds.take(1):
    print([x.shape for x in e])
te = time.time()
print("Duration: {} s".format(te - ts))
```
Output:
```
TensorFlow 2.1.0
Reading 200 tfrecord files...
<ParallelMapDataset shapes: {features: (130,), label: (5,)}, types: {features: tf.float32, label: tf.float32}>
<FlatMapDataset shapes: ((128,), (5,)), types: (tf.float32, tf.float32)>
<TakeDataset shapes: ((128,), (5,)), types: (tf.float32, tf.float32)>
<WindowDataset shapes: (DatasetSpec(TensorSpec(shape=(128,), dtype=tf.float32, name=None), TensorShape([])), DatasetSpec(TensorSpec(shape=(5,), dtype=tf.float32, name=None), TensorShape([]))), types: (DatasetSpec(TensorSpec(shape=(128,), dtype=tf.float32, name=None), TensorShape([])), DatasetSpec(TensorSpec(shape=(5,), dtype=tf.float32, name=None), TensorShape([])))>
<FlatMapDataset shapes: ((None, 128), (None, 5)), types: (tf.float32, tf.float32)>
<BatchDataset shapes: ((128, None, 128), (128, None, 5)), types: (tf.float32, tf.float32)>
<FlatMapDataset shapes: ((128, None, 128), (128, None, 5)), types: (tf.float32, tf.float32)>
Duration: 546.8976650238037 s
```

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.1.0 ,2.2.0-dev20200218 and was able to reproduce the issue.Please,find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/e06d28308eab617431c537750e6f6c5d/untitled671.ipynb). Thanks!

**Comment by mimxrt** [comment]
Thank you for verifiying this. I just found out that it seems to be caused by the `shift` parameter in the `.window()` function. It seems the larger the shift value, the slower it gets. For larger values it doesn't work at all any more.

**Comment by jsimsa** [comment]
TLDR: Your expectation that shape of the output data should not affect the performance is not grounded is reality. tf.data is a streaming API and the time it takes for "batch" transformations such as `window` and `batch` to produce an output is proportional to the "batch" dimension.

Both `window` and `batch` end up increasing the number of input elements needed to produce an output by a constant -- window size and batch size respectively. For instance, if it takes 1s to fetch a single output from `ds`, then -- assuming there is not asynchrony and parallelism within `ds` -- fetching a single output from `ds.window(size=10)` would be expected to take 10 seconds and fetching a single output from `ds.window(size=10).batch(batch_size=16)` would be expected to take 160 seconds. This is expected.

Increasing `shift` beyond `size` means that your input pipeline will need to read even more input elements per output. For example, to produce a single output from `ds.window(size=10, shift=20)`, you will need to consume 20 input elements per a single output element.

I recommend taking a look at the tf.data performance analysis [guide](https://www.tensorflow.org/guide/data_performance_analysis) to help you understand the performance of your input pipeline.

**Comment by mimxrt** [comment]
Thank you for your response. I think or rather hope there is a misunderstanding. What I did to compensate the difference in dimensionality and the resulting difference in total size is multiplicate the result time by the missing factor:

First, I stated the average size of the dimension 1 (number of time steps)

> Timings for comparison (note: number of time steps / dim 1 (None) is avg. **312.5** = 40000 / 128):

and consequently used it for the timing calculations.

> - 1x Shape ((128, None, 128), (128, None, 5)):
>  546.8976650238037 s
> - 1000x Shape ((128, 128), (128, 5)) (no windowing):
>  8.29618239402771 s * **312.5** / 1000 = 2.592556998133659375 s
>  Speedup: 210x faster
> - 1000x Shape ((None, 128), (None, 5)) (no batching):
>  65.43172574043274 * **128** / 1000 = 8,37526089477539072 s
>  Speedup: 65x faster

The first value of 546.9 seconds ist the reference with all dimensions.

The second value of 2.59 seconds is based on the benchmark without the "time step"-dimension taking 8.3 seconds. This "raw" value is multiplied by the average size of the missing dimension (312.5) and divided by the sample size 1000 (i.e. the loop ran 1000 rounds) to get to the final 2.59 second result.

Correspondingly, the third and final value of 8.38 seconds is computed with dimension 0 removed (batch size) and therefore, the "raw" value of 65.43 seconds is multiplied with the previously used batch size 128 for compensation. Again, the value is divided by 1000 (loop bound) which results in the final value.

Maybe I am missing something here - I hope you can help me understand my mistake.

I do understand, however, that the shift size can cause such behaviour. I maybe expected to see some more advanced caching mechanisms (after all the overlap of two windows contains the same data), but meanwhile I have found ways to pre-process the data differently and work around the issues described here.

**Comment by jsimsa** [comment]
Sorry, I didn't realize you are normalizing the runtime by the missing shape. I think I found the problem, which is that your input pipeline with both `batch` and `window`, does not produce any elements. This is because each dataset will be truncated to `ds = ds.take(trunc)` elements. Let's say that this is 40k (which is the average). Your batch size is `128`, so you window size will be `40000 // 128 = 312`. So far so good. The problem is that you are using `shift = win_size // 2` and `win_size` is fixed to be 2048 (as opposed to be being relative to the window size). Because of this, each window will be taken at 2048 offset and to create a batch of 128 windows, you would need at least 128 * 2048 elements, but you only have 40k. Because you are using `batch` with `drop_remainder=True`, partial batches will be dropped.

In other words, the performance you are listing as the baseline (using both `window` + `batch`), is the time needed to process all input data (and then throw it all away). You can validate my hypothesis by changing the loop bound of the baseline to an arbitrarily large number, which will not change the E2E time. Another way to validate this is to to set `drop_remainder=False`, which will result in only one partial batch being computed.

**Comment by mimxrt** [comment]
Yes, that makes sense, I should have thought of that. Thank you for taking the time and figuring this out!
